### ISSUE TITLE
Use a path to either a package's manifest (Cargo.toml) or a Wix source file (wxs)

### ISSUE THREAD

**Issue by volks73** [volks73/cargo-wix#70]
The create/default subcommand takes an optional input to a WiX source file (wxs). If no path is specified, then the default is to use the `wix\main.wxs` file from the current working directory (cwd) if it exists. All other subcommands take an optional input argument to a package's manifest (Cargo.toml), and similarly defaults to the cwd if no input is specified. So, all but the create/default command use the Cargo.toml file. 

This is inconsistent and could easily lead to usage errors. Instead, if the optional input argument is used with the create/default subcommand, then the file extension is used. If path is to a Cargo.toml file, it look for any WiX source file in the same folder as the Cargo.toml file and then for the `wix\main.wxs` relative to the Cargo.toml file. If the path is to a `wxs` file, then that is used for creation.

**Comment by volks73** [comment]
Another implement could be to always use the Cargo.toml file for all subcommands, including the create/default subcommand, but include a `-S,--src` or `-w,--wxs` option for the create/default subcommand that is used to a specify a WiX source that is not located in the default `wix` folder location.

**Comment by zacps** [comment]
I just started experimenting with this for Alacritty, this is looking pretty good!

It would be most convenient if we could specify the path to the `wxs` file in `Cargo.toml` and then have that used unless a different one is explicitly specified.

**Comment by volks73** [comment]
>It would be most convenient if we could specify the path to the wxs file in Cargo.toml and then have that used unless a different one is explicitly specified.

Yes, good idea. This is closely related to #20, where a Wix section will be added to the Cargo.toml file. One of the fields/keys for the eventual Wix section could be `src` or `source`, which specifies a path to the WXS file for the project. Any CLI options would override this value, but this value would override the hard-coded default, i.e. `wix\main.wxs`.

**Comment by volks73** [comment]
>One of the fields/keys for the eventual Wix section could be `src` or `source`, which specifies a path to the WXS file for the project. Any CLI options would override this value, but this value would override the hard-coded default, i.e. wix\main.wxs.

I decided to use `input` so that it would match with the `output` field and be consistent with the CLI argument naming as well.

**Comment by volks73** [comment]
I started working on this a little and ran into some problems. This is probably going to be more work than expected. The main problem is what to do with the build artifacts (.wixobj) if a Cargo.toml file is specified. Where should these files go, then what about the output/destination? If the subcommand is called from outside a Cargo project, or worse, from another Cargo project, where should all of these files go and should the output/destination be determined from the input or the output option be required? This all makes it more complicated.

With the implementation of #20, I do not think this is a as important. Instead, the path to a WXS file can be specified with the `input` field for the metadata section. I believe the focus should still be on creating installers for a cargo-based project from within the folder structure of the project.

**Comment by Songtronix** [comment]
Is there a way to use multiple wxs files? It looks like they won't get linked and I can't specify it anywhere.

**Comment by volks73** [comment]
Currently, no. The subcommand is written to only handle a single WXS file, but this should probably be a feature. I have created #92 to track adding this feature.

**Comment by volks73** [comment]
With the implementation of #92, I am not sure how this would work anymore. Could multiple Cargo.toml files be used as input? If so, then where are the output and object files located? Maybe there would have to be some validation such that only one Cargo.toml file is allowed? This really complicates things.

Instead of complicating the `<INPUTS>` argument for the default _create_ subcommand, maybe specifying a source, origin, manifest file, etc. should be an option. Something like, `-P,--project <MANIFEST>`, `-s,--src <MANIFEST>`, `-m,--manifest <MANIFEST>`, etc. I am currently leaning towards the `-M,--manifest` option since it is mnemonic and follows the convention of less used/"unsafe" options starting with capital letters. However, this is still inconsistent with the other subcommands that all take a single positional, optional `<INPUT>` argument for a package's manifest. Note, Cargo has the `--manifest-path <PATH>` option.

To be consistent with the other subcommands, the new `<INPUTS>` argument could be changed back to a single `<INPUT>` value that takes a Cargo.toml file, but multiple WXS files can be specified with multiple `-w,--wxs <WXS>` options in addition to any WXS files found in the default `wix` location relative to the location of the specified Cargo.toml file. Output would be relative to the specified Cargo.toml file in the `target\wix` folder, along with the wixobj files, unless the `-o,--output` option was used for the MSI file. Take the following project organization:

```dos
|-- C:\Path\to\Project
|-- |-- Cargo.toml
|-- |-- README.md
|-- |-- src
|-- |-- wix
|-- |-- |-- one.wxs
|-- |-- |-- two.wxs
|-- C:\Path\to\Common\Wxs\Files
|-- |-- three.wxs
|-- |-- four.wxs
```

The invocation of creating the installer for this example project where the CWD is actually `C:\` would be:

```dos
C:\> cargo wix --wxs "C:\Path\to\Common\Wxs\Files\three.wxs" --wxs "C:\Path\to\Common\Wxs\Files\four.wxs" "C:\Path\to\Project\Cargo.toml"
```

By default as of #92, `one.wxs` and `two.wxs` are automatically included in creating the installer. The `inputs` field for the `[package.metadata.wix]` would need to be changed to `wxs`, `sources`, `include`, `src`. I am kind of leaning towards `include` and maybe having the CLI option be `-I,--include <WXS>` since `-i` is already taken with `-i,--install-version`, but this might imply there is an exclude option of some kind. Maybe `-I,--import` and an `import` field is more consistent with function? Note, the `-i,--install-version` could not be renamed because there is already the `-V,--version` flag for the subcommand's version. The `source` variants might be confusing with the `src` folder for the Rust code.

I think this implementation might even work for workspaces (#74) when creating installers for sub-packages instead of needing to implement the `-p,--package <NAME>` option (#94):

```dos
C:\Path\to\Workspace\Project> cargo wix "client\Cargo.toml"
```

which would use the client's Cargo.toml file that includes the package fields necessary to create the installer and/or the `[package.metadata.wix]` section. Output would be in `C:\Path\to\Workspace\Project\client\target\wix` and WXS files would be included from `C:\Path\to\Workspace\Project\client\wix`. This does not resolve the all-in-one workspace installer enhancement, but I am still not sure that is even needed.

How useful is this enhancement anyways? It would be helpful for testing because then instead of changing the cwd for each test and requiring the `--test-threads=1` option, a path to a Cargo.toml file could be passed to build the various paths.

**Comment by volks73** [comment]
This has been implemented as of 4cf7cc0584a6d7f5cad29cfaf1b7aa4c3432ee28.

I decided to go with the `-I,--include <WXS>` implementation mentioned in my previous comment and change the `<INPUTS>` argument to `<INPUT>`, which takes a single value that is a path to a `Cargo.toml` file. The `[package.metadata.wix]` section is also changed to use an `include` field that is a TOML array and does _not_ recognize the `input` or `inputs` fields. Information about this change has been added to the binaries documentation as well.

This implementation makes it possible to create an installer for a project while not being in project's root directory, i.e. the current working directory (cwd) can be anywhere. Tests were added and currently pass for this implementation. 

This also makes the default _create_ subcommand consistent with the other subcommands. In other words, all subcommands can take an option path to a package's manifest (Cargo.toml) as its first positional argument, with the exception of the _print_ subcommand, where a template is the first positional argument but the second positional argument is a path to a package's manifest.
